// @slickgrid-universal/common/dist/commonjs/enums/extensionName.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionName = void 0;
/** List of available SlickGrid Extensions (Controls & Plugins) */
var ExtensionName;
(function (ExtensionName) {
    ExtensionName["autoTooltip"] = "autoTooltip";
    ExtensionName["cellExternalCopyManager"] = "cellExternalCopyManager";
    ExtensionName["cellMenu"] = "cellMenu";
    ExtensionName["checkboxSelector"] = "checkboxSelector";
    ExtensionName["columnPicker"] = "columnPicker";
    ExtensionName["contextMenu"] = "contextMenu";
    ExtensionName["draggableGrouping"] = "draggableGrouping";
    ExtensionName["groupItemMetaProvider"] = "groupItemMetaProvider";
    ExtensionName["gridMenu"] = "gridMenu";
    ExtensionName["headerButton"] = "headerButton";
    ExtensionName["headerMenu"] = "headerMenu";
    ExtensionName["rowDetailView"] = "rowDetailView";
    ExtensionName["rowMoveManager"] = "rowMoveManager";
    ExtensionName["rowSelection"] = "rowSelection";
})(ExtensionName = exports.ExtensionName || (exports.ExtensionName = {}));
