// @slickgrid-universal/rxjs-observable/dist/commonjs/rxjs.resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RxJsResource = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
class RxJsResource {
    constructor() {
        this.className = 'RxJsResource';
    }
    /**
     * The same Observable instance returned by any call to without a scheduler.
     * This returns the EMPTY constant from RxJS
     */
    get EMPTY() {
        return rxjs_1.EMPTY;
    }
    /** Simple method to create an Observable */
    createObservable() {
        return new rxjs_1.Observable();
    }
    /** Simple method to create an Subject */
    createSubject() {
        return new rxjs_1.Subject();
    }
    firstValueFrom(source) {
        return rxjs_1.firstValueFrom(source);
    }
    iif(condition, trueResult, falseResult) {
        return rxjs_1.iif(condition, trueResult, falseResult);
    }
    /** Tests to see if the object is an RxJS Observable */
    isObservable(obj) {
        return rxjs_1.isObservable(obj);
    }
    /** Emits the values emitted by the source Observable until a `notifier` Observable emits a value. */
    takeUntil(notifier) {
        return operators_1.takeUntil(notifier);
    }
}
exports.RxJsResource = RxJsResource;
