// @slickgrid-universal/rxjs-observable/dist/esm/rxjs.resource.js
import { EMPTY, iif, isObservable, firstValueFrom, Observable, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
export class RxJsResource {
    constructor() {
        this.className = 'RxJsResource';
    }
    /**
     * The same Observable instance returned by any call to without a scheduler.
     * This returns the EMPTY constant from RxJS
     */
    get EMPTY() {
        return EMPTY;
    }
    /** Simple method to create an Observable */
    createObservable() {
        return new Observable();
    }
    /** Simple method to create an Subject */
    createSubject() {
        return new Subject();
    }
    firstValueFrom(source) {
        return firstValueFrom(source);
    }
    iif(condition, trueResult, falseResult) {
        return iif(condition, trueResult, falseResult);
    }
    /** Tests to see if the object is an RxJS Observable */
    isObservable(obj) {
        return isObservable(obj);
    }
    /** Emits the values emitted by the source Observable until a `notifier` Observable emits a value. */
    takeUntil(notifier) {
        return takeUntil(notifier);
    }
}
